### DISCUSSION TITLE
Regarding python subprocess and venv in gramine

### DISCUSSION THREAD

**Discussion by chanfm81** [gramineproject/gramine#1554] [General]
I have the following code which runs a subprocess from main python process. 
The child process is using a different virtual environment which is created and installed prior to gramine runtime.
`    cmd = '. env/bin/activate && env/bin/python numpy_test.py && deactivate'`
`result = subprocess.check_output(cmd, shell=True)`

I cant find any discussion related to using python virtual environment in gramine.

Anyone has a different way of doing it?

Thanks

**Comment by dimakuv** [comment]
@chanfm81 Unfortunately, what you're trying to do is too advanced for Gramine. I'm afraid Gramine won't be able to support such use case.

Also, why do you need two different virtual environments for Python?

**Comment by chanfm81** [comment]
@dimakuv The context is with regards to the case of federated learning where both client and server run inside the enlcave. 
The client acts as an orchestrator which is forever listening to tasks from server.
The data scientist will send the training script to the server where the server dispatches it to the client via https. 
The data scientist will specify the requirements.txt such that the client will install prior to running the script.
Hence the client needs to launch a sub process to execute the script within the  different virtual environment.

**Reply by dimakuv** [reply]
Actually, it feels like the client and the server should run in **separate independent** SGX enclaves.

Anyway, what you're currently trying to do is way out of capabilities of Gramine-SGX. I would suggest to re-think your architecture.

**Reply by chanfm81** [reply]
Both the server and client are running on different sgx machines hence they are separate and independent.
The issue is on the client.
How can the client build a new environment and launch another enclave and pass its code from its own enclave?

**Reply by dimakuv** [reply]
Oh I see now. I incorrectly understood your scenario.

> How can the client build a new environment and launch another enclave and pass its code from its own enclave?

Could you split this functionality into two parts: the "building-venv-client" which will run in its own SGX enclave, and the "running-actual-script-client" which will run in a separate SGX enclave?

In fact, maybe you don't even need to first client to run in an SGX enclave? If the only thing that this client does is to prepare venv, why it's required to be in the SGX enclave?

The script(s) that the data scientist sends can be in encrypted form, and only the "running-actual-script-client" enclave should decrypt them (by getting a correct encryption key from the data scientist).

In other words, I propose to have a clear split:
1. The "preliminary" logic which is *not running in SGX enclave* and simply prepares the venv.
2. The "actual script" logic which runs in the Gramine SGX enclave, gets the encrypted script, decrypts it and runs it.

The data scientist coordinates (remotely) both these steps.

**Reply by chanfm81** [reply]
ok i will consider the above steps. 

If the first client has to be in enclave, will this scenario be supported in future?

Thanks.

**Reply by dimakuv** [reply]
> If the first client has to be in enclave, will this scenario be supported in future?

Hard to say. The problem is not in any particular feature that is missing from Gramine. The problem is more theoretical -- you're trying to do some sysadmin actions (installing a venv) inside Gramine / Intel SGX. That's just something that is **not** considered by Gramine or Intel SGX technologies. It may work, but most probably it will never work.

**Reply by chanfm81** [reply]
its not installing a venv but loading an existing venv in enclave.

**Comment by lejunzhu** [comment]
Venv activate script only touches environment variables, so Gramine should able to support it. For example, if we create a venv and install a different version of numpy:
```
python -m venv venv
./venv/bin/pip install numpy
```
Then run this python script:
```
import sys, os, subprocess

print("Host")
print(sys.path)

import numpy
print(numpy.__version__)

print("Venv")

VENV='/venv'
env = os.environ.copy()
env['PATH'] = VENV+'/bin:'+env['PATH'] if 'PATH' in env else VENV+'/bin'
env['VIRTUAL_ENV'] = VENV
del env['PYTHONPATH'] # Or set the correct value for venv

ret = subprocess.run([VENV+'/bin/python', '-c', 'import sys; print(sys.path); import numpy; print(numpy.__version__)'], capture_output=True, env=env)
print(ret.stdout.decode('utf-8'))
```
We'll see that the subprocess is using the newly installed package.

**Reply by dimakuv** [reply]
Thanks @lejunzhu for the technical details on venv!

If it's only the environment variables, then Python venv is a much simpler technology than I expected :)

In this case, I take my words back. Gramine should be able to work with this. The only problematic part is to enable the correct paths in the Gramine manifest.

**Comment by chanfm81** [comment]
Thank you all for the comments.

The code by @lejunzhu got the following error:
"Error occurred during checks: KeyError('PYTHONPATH')"

So I went to set PYTHONPATH instead of deleting it and it works.

My original code also works using shell=True in subprocess.

Cheers :)
